### DISCUSSION TITLE
Detecting if a player has an advancement and returning a boolean.

### DISCUSSION THREAD

**Discussion by Desynq** [FabricMC/fabric-api#2382] [Mod Dev Support] [answered]
I've been prowling through the Minecraft source code using the yarn mappings and I can't find any getter method that looks like:
```java
public static boolean hasAdvancement(PlayerEntity player, Advancement advancement) {
    ...
}
```
Does anyone know of any library that adds this, how to implement it myself, or of any pre-existing method similar to this?

**Comment by apple502j** [comment, marked as answer]
Of course there is. Any functionality available in vanilla Minecraft would have an implementation somewhere. That's how coding works.

As you are probably aware, static methods are rare - because Java has a better way to represent those. (OOP, as people call it.) It's probably named `has` or `get` or `isDone` or something like that, on the player or a class related to advancements.

Let's start from one thing we know: `ServerPlayerEntity`. (I assume serverside here, clients need to request advancement through packet.) There are two methos with the name "advancement" in it. One of them is clearly not the one we want, so we investigate the other one, `getAdvancementTracker`.

We reach `PlayerAdvancementTracker`. There are many methods and this class is undocumented. One of the ways to filter this is to ignore private methods and check public methods first. Also, we can ignore something that has void return type or is definitely a setter, like `grantCriterion`. There is only one method left, `getProgress`. This takes an `Advancement`. Let's check the return value. In theory there should be no value left to pass. We started with a player and we passed an advancement. Any method with an argument are therefore unlikely to be the one we want from this point.

There it is. `boolean isDone()`, which presumably returns true if the advancement progress is done. That's it. (What about `isAnyObtained`? Yes, that could be an option. But if you check the document, you'll notice that an advancement can have multiple criteria, and there exist a thing called unobtained criteria. In that case, `isAnyObtained` probably returns true if any of the criteria are obtained - but the advancement requires all of them. Otherwise why do people care about unobtained criteria?)


TL;DR: `serverPlayerEntity.getAdvancementTracker().getProgress(advancement).isDone()`

**Reply by Desynq** [reply]
Thanks for the in-depth response, it really helped with understanding the reasoning behind why it's done the way it is.
